src/plan_computation.py

```diff
@@ -110,4 +114,119 @@ def generate_photo_plan_on_grid(camera: Camera, dataset_spec: DatasetSpec) -> T.
         List[Waypoint]: scan plan as a list of waypoints.
 
     """
-    raise NotImplementedError()
+    scan_dimension_x = dataset_spec.scan_dimension_x
+    scan_dimension_y = dataset_spec.scan_dimension_y
+
+    max_computed_distances = compute_distance_between_images_with_angle(camera, dataset_spec,
+                                                                            angle_deg_x,
+                                                                            angle_deg_y)
+
+    # assume allowed_movment_px = 1
+    computed_speed = compute_speed_during_photo_capture(camera, dataset_spec, allowed_movement_px=1)
+
+    # calculate the number of photos in each direction
+    # we need at least two waypoints in each direction at the start and end point of the scan area
+    num_x = 1 + math.ceil(scan_dimension_x / max_computed_distances[0])
+    num_y = 1 + math.ceil(scan_dimension_y / max_computed_distances[1])
+
+    # actual distance between photos, based on actual number of photos to take
+    distance_between_photos_x = scan_dimension_x / (num_x - 1)
+    distance_between_photos_y = scan_dimension_y / (num_y - 1)
+
+    # Calculate adjustments of waypoints in x and y directions based on angles
+    adj_x = dataset_spec.height * math.tan(math.radians(angle_deg_x))
+    adj_y = dataset_spec.height * math.tan(math.radians(angle_deg_y))
+
+    result: T.List[Waypoint] = []
+    for j in range(0, num_y):
+        y = j * distance_between_photos_y # y position of camera
+        y -= adj_y # adjust y position based on camera gimbal angle y
+        for i in range(0, num_x):
+            # use lawnmower pattern, so x travels in reverse direction if j is odd
+            x = i * distance_between_photos_x if j%2==0 else (num_x - 1 - i) * distance_between_photos_x
+            x -= adj_x # adjust x position based on camera gimbal angle x
+            waypoint = Waypoint(x=x,y=y,
+                                speed=computed_speed, 
+                                camera_angle_x_deg=angle_deg_x, camera_angle_y_deg=angle_deg_y)
+            result.append(waypoint)
+
+    return result
+
+
+def time_between_waypoints(wp1: Waypoint, wp2: Waypoint, max_speed: float, max_acc: float) -> float:
```

Comment:
Nice work on this function: fixing my mistake as well as detailed explanations and derivation!